src/core/models/outputs/outputs.c

```diff
@@ -3,80 +3,313 @@
 
 #include <inttypes.h>
 #include <stdio.h>
+#include <stdlib.h>
 
+#include "core/models/outputs/output_alias.h"
+#include "core/models/outputs/output_extended.h"
+#include "core/models/outputs/output_foundry.h"
+#include "core/models/outputs/output_nft.h"
 #include "core/models/outputs/outputs.h"
+#include "utlist.h"
 
-#define UTXO_OUTPUT_MIN_COUNT 0
-#define UTXO_OUTPUT_MAX_COUNT 126
+#define UTXO_OUTPUT_MAX_COUNT 127
 
-int utxo_outputs_add(outputs_ht **ht, output_type_t type, byte_t addr[], uint64_t amount) {
-  if (type == OUTPUT_DUST_ALLOWANCE && amount < 1000000) {
-    printf("[%s:%d] dust allowance amount must at least 1Mi\n", __func__, __LINE__);
-    return -1;
+utxo_outputs_list_t *utxo_outputs_new() { return NULL; }
+
+void utxo_outputs_free(utxo_outputs_list_t *outputs) {
+  if (outputs) {
+    utxo_outputs_list_t *elm, *tmp;
+    LL_FOREACH_SAFE(outputs, elm, tmp) {
+      switch (elm->output->output_type) {
+        case OUTPUT_EXTENDED:
+          output_extended_free(elm->output->output);
+          break;
+        case OUTPUT_ALIAS:
+          output_alias_free(elm->output->output);
+          break;
+        case OUTPUT_FOUNDRY:
+          output_foundry_free(elm->output->output);
+          break;
+        case OUTPUT_NFT:
+          output_nft_free(elm->output->output);
+          break;
+      }
+      free(elm->output);
+      LL_DELETE(outputs, elm);
+      free(elm);
+    }
   }
+}
 
-  if (utxo_outputs_count(ht) >= UTXO_OUTPUT_MAX_COUNT) {
-    printf("[%s:%d] output count must be < 127\n", __func__, __LINE__);
+int utxo_outputs_add(utxo_outputs_list_t **outputs, utxo_output_type_t type, void *output) {
+  if (output == NULL) {
+    printf("[%s:%d] invalid parameters\n", __func__, __LINE__);
     return -1;
   }
 
-  outputs_ht *elm = utxo_outputs_find_by_addr(ht, addr);
-  if (elm) {
-    printf("[%s:%d] address exists\n", __func__, __LINE__);
+  if (utxo_outputs_count(*outputs) >= UTXO_OUTPUT_MAX_COUNT) {
+    printf("[%s:%d] output count must be <= 127\n", __func__, __LINE__);
     return -1;
   }
 
-  elm = malloc(sizeof(outputs_ht));
-  if (elm == NULL) {
-    printf("[%s:%d] OOM\n", __func__, __LINE__);
-    return -1;
+  utxo_outputs_list_t *next = malloc(sizeof(utxo_outputs_list_t));
+  if (next) {
+    next->output = malloc(sizeof(utxo_output_t));
+    if (next->output) {
+      next->output->output_type = type;
+      switch (next->output->output_type) {
+        case OUTPUT_EXTENDED:
+          next->output->output = malloc(sizeof(output_extended_t));
+          memcpy(next->output->output, output, sizeof(output_extended_t));
```

Comment:
simply memcpy will lead to potential memory issues. 
for instance, users supposed to use `output_extended_new`/`output_extended_free` but `output_extended_free` will cause double-free, say if you call `output_extended_free(extended_output)` in test_outputs.c.

Two ways for preventing memory leak and better memory management,
1) simply gives the pointer of the output to the list 
```
next->output->output = output;
```
the list will need to free the output object but the caller doesn't need to do that.
it reduces memory usage but might have problems if the caller reuses the pointer or free it somewhere.

2) duplicated the output object in this add method, see `address_clone`
```
next->output->output = output_extended_clone(output);
```
the caller should free the output via `output_extended_free`.
it consumes more memory but can guarantee the object is managed by the list.

I prefer the 2nd approach, what do you think?

Reply:
I see where is the problem:
`utxo_outputs_list_t` list shares `output_extended_t` structure with a caller of `utxo_outputs_add` function if `memcpy` is used because `memcpy` only copies a pointers which are in `output_extended_t` structure.
Rather a whole new `output_extended_t` structure needs to be created for `utxo_outputs_list_t` list.
The same as stated above is true for other output types as well.

Yes, 2nd approach seems to be better choice and more aligned with other modules which also create a new instance of an object if necessary.

Reply:
Done